core/primitives/src/crypto/signer.rs

```diff
@@ -198,36 +204,62 @@ impl InMemorySigner {
         let mut rng = OsRng::new().expect("Unable to generate random numbers");
         let account_id: String =
             rng.sample_iter(&Alphanumeric).filter(|c| c.is_ascii_alphabetic()).take(10).collect();
-        let (public_key, secret_key) = signature::get_key_pair();
+        let (public_key, secret_key) = get_key_pair();
         let bls_secret_key = BlsSecretKey::generate();
         let bls_public_key = bls_secret_key.get_public_key();
         Self { account_id, public_key, secret_key, bls_public_key, bls_secret_key }
     }
 }
 
+// Use hash for the Key so that we do not store raw crypto keys in the memory. Also more memory
+// efficient.
+cached_key!{
+  TRANSACTION_CACHE: SizedCache<CryptoHash, Signature> = SizedCache::with_size(1_000_000);
+  Key = {
+        let mut res = sk.to_bytes();
+        res.extend_from_slice(data);
+        hash(&res)
+        };
+  fn sign_transaction(data: &[u8], sk: &SecretKey) -> Signature  = {
+    sign(data, sk)
+  }
+}
+
 impl TransactionSigner for InMemorySigner {
     #[inline]
     fn public_key(&self) -> PublicKey {
         self.public_key
     }
 
-    fn sign(&self, data: &[u8]) -> signature::Signature {
-        signature::sign(data, &self.secret_key)
+    fn sign(&self, data: &[u8]) -> Signature {
+        sign_transaction(data, &self.secret_key)
     }
 }
 
+cached_key!{
+  BLOCK_CACHE: SizedCache<CryptoHash, PartialSignature> = SizedCache::with_size(100_000);
```

Comment:
100_000 is too much, this is just for the block for which we're aggregating bls signatures?
Nightshade also uses BLS signatures, but it calls bls_sign and seems like it wouldn't benefit from caching

Also, we can add some debug counters for cache hits and cache misses

Reply:
> 100_000 is too much
Our BLS signature is 320bits, correct? 100k BLS signatures is 4MiB of RAM. This is really not much.
> this is just for the block for which we're aggregating bls signatures?
Also PayloadGossip. Suppose we have 1k validators on a beacon chain. Each of them provides 3 bls signatures: payload gossip + two signatures for the block. 100k would cache 30 blocks, we could reduce it to 10 blocks, but it is 4MiB or RAM vs 1MiB not a drastic change. Also unit tests that launch multiple nodes would share that same LRU (as a consequence of it being static). 

I agree that we need a more data-driven approach for determining the size of the cache, this would require writing benchmarks. Also it is a part of a larger body of work, see: https://github.com/nearprotocol/nearcore/issues/833
I also do not want to block #829 (which requires this PR). Let's address it later.

Reply:
this cache is only for your own signature, not other validators

Reply:
Actually you are correct, but again it is 4MiB we are debating over which is something really small compared to our general footprint. If we want to have a proper cache size evaluation then it should be done separately from this PR as discussed in #833.

Reply:
It should be possible to make our own cached! macro using SizedCache that
- stores uses the cache stored in the struct, caches a method on &self
- maintains a counter for cache hits and misses
- prints log warnings if we get more misses than hits

Reply:
Sure. That's a lot of work, let's address it in #833. I do not want us to be blocked.

Reply:
Turns out cache hits and misses are already counted:
https://docs.rs/cached/0.8.0/cached/trait.Cached.html
We don't have to make a macro, but can we
- put a SizedCache into InMemorySigner
- make bls_sign_with_cache do a cache_get and cache_set if it fails

I don't like this cache living outside of Signer and hashing the secret key to work around that

Reply:
What hits and misses will be used for?

Reply:
don't need to be used, but good to know the library has them

Reply:
> prints log warnings if we get more misses than hits
I suggest we do not logging info just in case until it is actually used by someone. We have a proliferation of logging in our code. Also I do not think log crate supports logging every Nth time.

Reply:
Fixed.